### PR TITLE
blockdev-util: introduce block_device_get_whole_disk()

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

File: src/shared/blockdev-util.c [hunk 1]
```diff
@@ -518,19 +538,12 @@ int partition_enumerator_new(sd_device *dev, sd_device_enumerator **ret) {
...
-                return -ENOTBLK;
-
-        r = sd_device_get_devtype(dev, &s);
-        if (r < 0)
-                return r;
-
-        if (!streq(s, "disk")) /* Refuse invocation on partition block device, insist on "whole" device */
-                return -EINVAL;
+        if (r == 0)
+                return -ENXIO;
```
Comment (daandemeyer) [inline src/shared/blockdev-util.c]: Shouldn't this stay EINVAL if we want no functional changes?

File: src/shared/blockdev-util.c [hunk 2]
```diff
@@ -20,6 +20,26 @@
...
+        const char *s;
+        int r;
+
+        assert(dev);
+
+        r = sd_device_get_subsystem(dev, &s);
+        if (r < 0)
+                return r;
+
+        if (streq(s, "block"))
```
Comment (daandemeyer) [inline src/shared/blockdev-util.c]: Missing `!` here?

File: src/udev/udevd.c [hunk 3]
```diff
@@ -493,20 +486,15 @@ static int device_get_whole_disk(sd_device *dev, sd_device **ret_device, const c
...
-        if (r >= 0 && streq(val, "partition")) {
-                r = sd_device_get_parent(dev, &dev);
-                if (r == -ENOENT) /* The device may be already removed. */
-                        goto irrelevant;
-                if (r < 0)
-                        return log_device_debug_errno(dev, r, "Failed to get parent device: %m");
-        }
+        r = block_device_get_whole_disk(dev, &dev);
+        if (IN_SET(r,
+                   -ENOTBLK, /* The device is not block device. */
```
Comment (daandemeyer) [inline src/udev/udevd.c]: "is not a block device"

File: src/udev/udevd.c [hunk 4]
```diff
@@ -493,20 +486,15 @@ static int device_get_whole_disk(sd_device *dev, sd_device **ret_device, const c
...
+                   -ENODEV   /* The whole disk device not found, may be already removed. */))
```
Comment (daandemeyer) [inline src/udev/udevd.c]: "was not found, it may already be removed"

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: src/shared/blockdev-util.c (hunk 2 above)
Comment (yuwata) [reply src/shared/blockdev-util.c]: Ouch!

**Comment by yuwata** [comment]
@DaanDeMeyer Thank you for the review. Force-pushed a revised version. Now, the first commit certainly not change the behavior, including the return value. And changing the return value is moved to another commit. PTAL.

File: src/udev/udevd.c [hunk 5]
```diff
@@ -493,20 +486,15 @@ static int device_get_whole_disk(sd_device *dev, sd_device **ret_device, const c
...
+        if (IN_SET(r,
+                   -ENOTBLK, /* The device is not a block device. */
+                   -ENODEV   /* The whole disk device was not found, it may already be removed. */))
+                goto irrelevant;
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to get whole disk device: %m");
 
         r = sd_device_get_devname(dev, &val);
-        if (r == -ENOENT)
-                goto irrelevant;
```
Comment (daandemeyer) [inline src/udev/udevd.c:509]: Why drop this? This seems orthogonal to using block_device_get_whole_disk()

**Review by DaanDeMeyer — APPROVED** [review]

File: src/udev/udevd.c (hunk 5 above)
Comment (yuwata) [reply src/udev/udevd.c:509]: Haha, yeah, I will split the change and explain why I did.

The failure should actually not occur. We already confirmed that the device is a whole block device in the above, hence the device should have a corresponding device node. So, I dropped the condition.

**Comment by yuwata** [comment]
@DaanDeMeyer Thank you for the review. The last commit was split into two, and explained why in the commit message. Upgrading the green label.
